org.eclipse.jdt.core/model/org/eclipse/jdt/internal/compiler/CompilerConfiguration.java

```diff
@@ -0,0 +1,67 @@
+/*******************************************************************************
+* Copyright (c) 2024 Microsoft Corporation and others.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Eclipse Public License 2.0
+* which accompanies this distribution, and is available at
+* https://www.eclipse.org/legal/epl-2.0/
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Contributors:
+*     Microsoft Corporation - initial API and implementation
+*******************************************************************************/
+
+package org.eclipse.jdt.internal.compiler;
+
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.core.resources.IContainer;
+
+/**
+ * This class encapsulates the standard compiler options that can be
+ * used to compile Java files. It provides methods to set and retrieve
+ * various compiler options, including source paths, class paths,
+ * output directories, annotation processing options, and other compiler
+ * arguments.
+ *
+ * Clients typically use this class when opting for an alternative compiler
+ * like Javac to compile Java files.
+ *
+ * @since 3.38
+ */
+public final record CompilerConfiguration(
+		/**
+		 * List of file paths where the compiler can find source files.
+		 */
+		List<IContainer> sourcepaths,
+		/**
+		 * List of file paths where the compiler can find source files for modules.
+		 */
+		List<IContainer> moduleSourcepaths,
+		/**
+		 * List of file paths where the compiler can find user class files and annotation processors.
+		 */
+		List<Either<IContainer, String>> classpaths,
```

Comment:
I dislike the "Either" class. But i have not solution in hand :-(. I guess this is caused by the complicated definition of `org.eclipse.jdt.core.IClasspathEntry.getPath()` ? Can you instead pass a List of IClasspathEntry which is complicated but well defined?

Reply:
They are inferred from NameEnvironment of JavaBuilder, where it defines multi kinds of o.e.j.i.c.builder.ClasspathLocation . The problem for ClasspathLocation is that its property modifier are package visible.

Reply:
They are inferred from NameEnvironment of JavaBuilder, where it defines multi kinds of o.e.j.i.c.builder.ClasspathLocation . The problem for ClasspathLocation is that its property modifier are package visible.

Reply:
How about using URI?
Resource.getLocationURI() / java.io.File.toURI()

Reply:
URI is more general than path, and it's good for me to use URI instead.